doc/man7/openssl-core.h.pod

```diff
@@ -0,0 +1,251 @@
+function must set this field to indicate the actual size of the data
+stored in memory pointed at by the C<buffer> pointer, or the needed
+buffer size if the size indicated by C<buffer_size> is too small.
+
+=item C<data_type>
+
+=for comment It's still debated if this field should be present, or if
+the type should always be implied by how it's used.  Either way, these
+data types will have to be passed together with the names as an array
+of OSSL_ITEM, for discovery purposes.
+
+The C<data_type> is a value that describes the type and organization of
+the data, and can be one of:
+
+=over 4
+
+=item C<OSSL_PARAM_INTEGER>
+
+=item C<OSSL_PARAM_UNSIGNED_INTEGER>
+
+The parameter data is an integer (signed or unsigned) of arbitrary
+length, organized in native form, i.e. most significant byte first on
+Big-Endian systems, and least significant byte first on Little-Endian
+systems.
+
+=item C<OSSL_PARAM_REAL>
+
+=for comment It's still debated if we need this or not.
+
+The parameter data is a floating point value in native form.
+
+=item C<OSSL_PARAM_UTF8_STRING>
+
+The parameter data is a printable string.
+
+The pointer form
+
+=item C<OSSL_PARAM_OCTET_STRING>
+
+The parameter data is an arbitrary string of bytes.
+
+=back
+
+Additionally, this flag can be added to any type:
+
+=over 4
+
+=item C<OSSL_PARAM_POINTER_FLAG>
+
+This can be used to indicate that constant data is or will be passed.
+It works by assuming that the memory pointed at by the C<buffer> field
+contains a pointer to the actual data.
+
+If this C<OSSL_PARAM> is used to set a parameter, C<buffer_size> must
+be set to the size of the data pointed at rather than the size of the
+memory pointed at by the C<buffer> field.
+If this C<OSSL_PARAM> is used in a parameter request, C<buffer_size>
+is not relevant.
+However, the responding function will set C<*return_size> to the size
+of the actual data.
+
+Note that the use of ths flag is B<fragile> and can only be safely
+used for data that remains constant and in a constant location for a
+long enough duration (such as the life-time of the object that
+provides these parameters).
+
+=back
+
+For convenience, these types are provided:
+
+=over 4
+
+=item C<OSSL_PARAM_UTF8_STRING_PTR>
+
+=item C<OSSL_PARAM_OCTET_STRING_PTR>
+
+These are combinations of C<OSSL_PARAM_UTF8_STRING> as well as
+C<OSSL_PARAM_OCTET_STRING> with C<OSSL_PARAM_POINTER_FLAG>.
+
+=back
+
+=back
+
+=back
+
+=head1 EXAMPLES
+
+=head2 OSSL_PARAM
+
+A couple of examples to just show how C<OSSL_PARAM> arrays could be
+set up.
+
+=head3 Example 1
+
+This example is for setting parameters on something:
+
+    #include <openssl/core.h>
+
+    const char *foo = "some string";
+    size_t foo_l = strlen(foo) + 1;
+    const char bar[] = "some other string";
+    const OSSL_PARAM set[] = {
+        { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, foo_l, NULL },
+        { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), NULL },
```

Comment:
So, now I'm confused in a different way. In the set case aren't we filling in a buffer provided to us be the requestor?

Reply:
No.  Argh terminology, please suggest something that works better for you.

I'll try to explain:

*Setting* parameters is when a caller initiates a call to some function to pass a set of parameters.  Say for example that you have an object reference to a newly created RSA key, and you want to set its diverse numbers (n, d, e, p, q, ...).  A way to do this is by setting up an OSSL_PARAM array with a number of OSSL_PARAM_UNSIGNED_INTEGER elements and pass that to whatever function that puts those number in the private bits of that key.

*Requesting* parameters is the other way around, a caller initiates a call to get the values of a set of parameters.  Say again that you hold a reference to an RSA key (that already have all its bits), and you want to find out its public numbers (n and e).  A way to do this is to set up an OSSL_PARAM array with the appropriate OSSL_PARAM_UNSIGNED_INTEGER elements, pointers to buffers that will receive those numbers and pointers to `size_t` variables to get the actual size of those numbers, and pass that to whatever function that then extracts those numbers from that key and fills in buffers and sizes.

Whether we talk about *setting* or *requesting*, it's from the perspective of the caller (the owner of the `OSSL_PARAM` array, if you wish).  *Responding* is from the perspective of the callee, though, and is what the function answering to a param request does.

Typically, I imagine that setting parameters would happen with a call to `{something}_set_params()`, while requesting parameters would happen with a call to `{something}_get_params()`

Does this clarify things?